wgengine/tsdns/tsdns.go

```diff
@@ -0,0 +1,115 @@
+// Copyright (c) 2020 Tailscale Inc & AUTHORS All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Package tsdns provides a Resolver struct capable of resolving
+// domains on a Taiscale network.
+package tsdns
+
+import (
+	"encoding/binary"
+	"errors"
+	"strconv"
+
+	"github.com/tailscale/wireguard-go/device"
+	"tailscale.com/types/logger"
+	"tailscale.com/wgengine/packet"
+)
+
+const (
+	// MaxQuerySize is the maximal size of a Magic DNS query.
+	MaxQuerySize = 512
+	// MaxResponseSize is the maximal size of a Magic DNS response.
+	MaxResponseSize = 512
+)
+
+const (
+	ipOffset      = device.MessageTransportHeaderSize
+	dnsDataOffset = ipOffset + packet.UDPDataOffset
+
+	bufferSize = dnsDataOffset + MaxResponseSize
+)
+
+// Resolver is a
+type Resolver struct {
+	logf logger.Logf
+
+	// ip is the IP on which the resolver is listening (default 100.100.100.100).
+	ip packet.IP
```

Comment:
for new code, try to use the inet.af/netaddr package where possible. (or let us know if it's not possible and we'll fix it)

Reply:
I've addressed most of the other feedback, but this seems difficult to fully eradicate from the codebase due to the way filters are currently written. I can refactor that eventually, but perhaps that's for a separate PR? It starts touching a significant number of things.